### DISCUSSION TITLE
How to include components when rendering on the server?

### DISCUSSION THREAD

**Discussion by unix** [vercel/next.js#10928] [Help] [answered]
**Qquestion description**:

  If I'm using a **third-party component** with a global style, All third-party components are bundle into **chunks**.

<br/>

**Example**:

  1. I have a component library [styled-jsx-lib-app](https://github.com/unix/styled-jsx-lib-app) (demo)

```jsx
const BaseCssComponent = () => {
  return (
    <div>
      <style global jsx>{`
        html {
          font-size: 30px;
        }
      `}</style>
    </div>
  )
}
```

  2. After running `next dev`, you can see that the **global style** is directly rendered:

![styled-jsx-lib-rendered](https://user-images.githubusercontent.com/11304944/76296371-047a8480-62f1-11ea-9d5b-a1a9b8c46db6.png)

  3. When I publish the library to NPM and use it in another project: [styled-jsx-user-app](https://github.com/unix/styled-jsx-user-app):

```
import { BaseCssComponent } from 'styled-jsx-lib-app'

const Home = () => (
  <div>
    hello, I am using library `styled-jsx-lib-app`.
    <BaseCssComponent />
  </div>
)
```
  4. This third-party library will always be in the chunk:

![styled-demo-2](https://user-images.githubusercontent.com/11304944/76296789-b4e88880-62f1-11ea-8774-dda21bd11a29.png)

<br/>

**I Want**:
I want to include this **base style component** when rendering on the server,  what should I do now?

<br/>

**Some of the solutions I've tried**:
  
  1. Publish `.css` file to npm, and import in the `_app.js` of `next.js`. This works, but I still want to publish a **component** if possible.

  2. Use `flush` from `styled-jsx/server` in `_document.jsx`. Not work, `flush` always return `[]`.

**Comment by unix** [comment, marked as answer]
I think I solved the problem.
Update results, may be helpful to others.
(🙏 can anyone help me close this discussion?)

`styled-jsx` cannot collect styles from `node_modules`, even if these **style components** are generated with `styled-jsx`.

As a solution, you can export the  `flush` of `styled-jsx/server` manually in the library, only in this way, we cat get [styleSheetRegistry](https://github.com/zeit/styled-jsx/blob/master/src/style.js#L4) instance.

Example:

In your component library:

```
import flush from 'styled-jsx/server'

const BaseCssComponent = () => {
  return (
    <div>
      <style global jsx>{`
        html {
          font-size: 30px;
        }
      `}</style>
    </div>
  )
}

BaseCssComponent.flush = flush
export default BaseCssComponent
// or export { ... flush }
```

// in next.js
```jsx
class MyDocument extends Document {
  static async getInitialProps (ctx) {
    const initialProps = await Document.getInitialProps(ctx)
    const styles = BaseCssComponent.flush()

    return {
      ...initialProps,
      styles: (
        <>
          {initialProps.styles}
          {styles}
        </>
      )
    }
  }

  ...
}
```

**Comment by rafaelalmeidatk** [comment]
I am not sure this is the reason, but is `styled-jsx` a peerDependency in your component library? Generally when building component libraries you always add the styling tools (SC, emotion, styled-jsx, etc.) as a peerDependency, because your project pick up the version the project that installs your library is using, avoiding this problem of having to reexport methods from your own library and duplicating things in the app

**Reply by unix** [reply]
`styled-jsx` is involved in packaging, and they are in the final bundle. (I mean it, because I need to publish a general package)

Anyway, I'm glad the problem has been solved, and thanks for your reply. 🎉

**Reply by rafaelalmeidatk** [reply]
Ah if you need it in the bundle then yeah it makes sense to not be a peerDependency, I'm glad you solved it.
